### ISSUE TITLE
NativeAOT: TypePreinit.ArrayInstance.TryStoreElement fails with invalid destination array size

### ISSUE THREAD

**Issue by christianscheuer** [dotnet/runtimelab#1185]
Hi,

We're trying to move from our old CoreRT-based build setup to most recent NativeAOT bits.

I'm getting this error in the midst of native code generation.
What's the best way to troubleshoot where this error comes from / help provide info?

This is a 2500+ file project, so narrowing it down manually might take a lot of time.

```
Destination array was not long enough. Check the destination index, length, and the array's lower bounds. (Parameter 'destinationArray') [/Users/chr/d/sf/pkg/sfbackend/src/sfbackend.csproj]
  System.ArgumentException: Destination array was not long enough. Check the destination index, length, and the array's lower bounds. (Parameter 'destinationArray')
     at System.Array.Copy(Array sourceArray, Int32 sourceIndex, Array destinationArray, Int32 destinationIndex, Int32 length, Boolean reliable)
     at System.Array.Copy(Array sourceArray, Int32 sourceIndex, Array destinationArray, Int32 destinationIndex, Int32 length)
     at ILCompiler.TypePreinit.ArrayInstance.TryStoreElement(Int32 index, Value value)
     at ILCompiler.TypePreinit.TryScanMethod(MethodIL methodIL, Value[] parameters, Stack`1 recursionProtect, Int32& instructionCounter, Value& returnValue)
     at ILCompiler.TypePreinit.ScanType(CompilationModuleGroup compilationGroup, ILProvider ilProvider, MetadataType type)
     at Internal.TypeSystem.LockFreeReaderHashtable`2.CreateValueAndEnsureValueIsInTable(TKey key)
     at ILCompiler.PreinitializationManager.IsPreinitialized(MetadataType type)
     at ILCompiler.DependencyAnalysis.NodeFactory.<CreateNodeCaches>b__39_3(MetadataType type)
     at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
     at ILCompiler.UsageBasedMetadataManager.GetRuntimeMappingDependenciesDueToReflectability(DependencyList& dependencies, NodeFactory factory, TypeDesc type)
     at ILCompiler.MetadataManager.GetDependenciesDueToReflectability(DependencyList& dependencies, NodeFactory factory, TypeDesc type)
     at ILCompiler.DependencyAnalysis.ConstructedEETypeNode.ComputeNonRelocationBasedDependencies(NodeFactory factory)
     at ILCompiler.DependencyAnalysis.ObjectNode.GetStaticDependencies(NodeFactory factory)
     at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.GetStaticDependenciesImpl(DependencyNodeCore`1 node)
     at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.GetStaticDependencies(DependencyNodeCore`1 node)
     at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.ProcessMarkStack()
     at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.ComputeMarkedNodes()
     at ILCompiler.ILScanner.ILCompiler.IILScanner.Scan()
     at ILCompiler.Program.Run(String[] args)
     at ILCompiler.Program.Main(String[] args)
/Users/chr/.nuget/packages/microsoft.dotnet.ilcompiler/6.0.0-preview.6.21278.2/build/Microsoft.NETCore.Native.targets(282,5): error MSB3073: The command ""/Users/chr/.nuget/packages/runtime.osx-x64.microsoft.dotnet.ilcompiler/6.0.0-preview.6.21278.2/tools/ilc" @"obj/Cloud/net5.0/osx-x64/native/sfbackend.ilc.rsp"" exited with code 1. [/Users/chr/d/sf/pkg/sfbackend/src/sfbackend.csproj]
```

**Comment by kant2002** [comment]
IMO the easiest way would be [just compile NativeAOT](https://github.com/dotnet/runtimelab/blob/feature/NativeAOT/docs/workflow/building/coreclr/nativeaot.md) and run locally build ILC under debugger. If you feed `obj/Cloud/net5.0/osx-x64/native/sfbackend.ilc.rsp` and catch exception, walking back to `ILCompiler.TypePreinit.TryScanMethod` should give you idea what method was failed to compile. or maybe `ILCompiler.TypePreinit.ScanType`.

You have a great chance that you unblock yourself and may create repro case.

**Comment by christianscheuer** [comment]
Thanks @kant2002. This appears to not be happening with the Debug build of ILC.

Do you know how to build the release mode of ILC?

When I run this, it produces directory in artifacts/packages/Debug/Shipping - not the Release version:

```
./build.sh nativeaot+libs+nativeaot.packages -rc Release -lc Release
```

**Comment by christianscheuer** [comment]
Turns out you need to pass `-c Release` as well to get the Release build:
```
./build.sh nativeaot+libs+nativeaot.packages -c Release -rc Release -lc Release
```

I feel like that could be made clearer here - let me know if you'd like a PR for this:
https://github.com/dotnet/runtimelab/blob/feature/NativeAOT/docs/workflow/building/coreclr/nativeaot.md#building

**Comment by christianscheuer** [comment]
I've made a repro case for this now here:
https://github.com/christianscheuer/runtimelab/tree/repro-scan-crash/repro

Turns out that ILC doesn't like to scan the static constructor of `ICSharpCode.SharpZipLib.Zip.Compression.InflaterHuffmanTree` from `SharpZipLib` version 1.2.0 (also tested with "*" for latest version, same issue).

This error appears to only happen when ILC is built with the `-c Release` switch, so not when built in Debug mode.

Code for the static constructor is here:
https://github.com/icsharpcode/SharpZipLib/blob/8c0a16904a0fa545c198c2ebe56335974925d120/src/ICSharpCode.SharpZipLib/Zip/Compression/InflaterHuffmanTree.cs#L34

**Comment by christianscheuer** [comment]
I've done some further reduction of the repro case. It appears it's related to reassigning an array with a new size:

```
using System;

class Program
{
    public static void Main()
    {
        Console.WriteLine("Repro");
    }

    static Program()
    {
        byte[] codeLengths = new byte[288];
        codeLengths = new byte[32];
        int i = 0;
        while (i < 32)
        {
            codeLengths[i++] = 5;
        }

    }
}
```

**Comment by MichalStrehovsky** [comment]
This is a bug in preinitialization - the component that interprets static constructors at compile time. If the static constructor can be interpreted, we don't compile it anymore and replace it with a data blob (that is the result of running the static constructor). This only runs if ILC was invoked with `-O` (to optimize) or `--preinitstatics`. It can be disabled by passing `--nopreinitistatics`.

We were lucky there was an off-by-one range check bug. Otherwise the failure mode would be more silent (either generating bad preinitialization data or aborting preinitialization - which would also be fine).

#1189 has the fix.

Thanks for reporting this and for a good repro!

**Comment by jkotas** [comment]
Fixed by #1189

**Comment by christianscheuer** [comment]
Thanks guys for the super speedy fix!
